login.go

```diff
@@ -40,6 +40,63 @@ func (c *LoginController) Authorize(ctx *app.AuthorizeLoginContext) error {
 	return c.auth.Perform(ctx)
 }
 
+// Refresh obtain a new access token using the refresh token.
+func (c *LoginController) Refresh(ctx *app.RefreshLoginContext) error {
+	refreshToken := ctx.Payload.RefreshToken
+	if refreshToken == nil {
+		return jsonapi.JSONErrorResponse(ctx, errors.NewBadParameterError("refresh_token", nil).Expected("not nil"))
+	}
+
+	client := &http.Client{Timeout: 10 * time.Second}
+	res, err := client.PostForm(configuration.GetKeycloakEndpointToken(), url.Values{
+		"client_id":     {configuration.GetKeycloakClientID()},
+		"client_secret": {configuration.GetKeycloakSecret()},
+		"refresh_token": {*refreshToken},
+		"grant_type":    {"refresh_token"},
+	})
+	if err != nil {
+		return jsonapi.JSONErrorResponse(ctx, errors.NewInternalError("Error when obtaining token "+err.Error()))
+	}
+	switch res.StatusCode {
+	case 200:
+		// OK
+	case 401:
+		return jsonapi.JSONErrorResponse(ctx, errors.NewUnauthorizedError(res.Status+" "+readBody(res.Body)))
+	case 400:
+		return jsonapi.JSONErrorResponse(ctx, errors.NewBadParameterError(readBody(res.Body), nil))
+	default:
+		return jsonapi.JSONErrorResponse(ctx, errors.NewInternalError(res.Status+""+readBody(res.Body)))
+	}
+
+	token, err := readToken(res, ctx)
+	if err != nil {
+		return err
+	}
+
+	return ctx.OK(&app.AuthToken{Token: token})
+}
+
+func readBody(body io.ReadCloser) string {
+	buf := new(bytes.Buffer)
+	buf.ReadFrom(body)
+	return buf.String()
+}
+
+func readToken(res *http.Response, ctx jsonapi.InternalServerError) (*app.TokenData, error) {
+	// Read the json out of the response body
+	buf := new(bytes.Buffer)
+	io.Copy(buf, res.Body)
+	res.Body.Close()
+	jsonString := strings.TrimSpace(buf.String())
+
+	var token app.TokenData
+	err := json.Unmarshal([]byte(jsonString), &token)
+	if err != nil {
+		return nil, jsonapi.JSONErrorResponse(ctx, errors.NewInternalError(fmt.Sprintf("Error when unmarshal json with access token %s ", jsonString)+err.Error()))
```

Comment:
As a future convention, we might want to avoid capitalized error messages when possible. https://github.com/golang/go/wiki/CodeReviewComments#error-strings